spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -887,7 +890,18 @@ protected void pollAndInvoke() {
 			}
 			checkPaused();
 			this.lastPoll = System.currentTimeMillis();
+			this.polling.set(true);
 			ConsumerRecords<K, V> records = this.consumer.poll(this.pollTimeout);
+			if (!this.polling.compareAndSet(true, false)) {
+				/*
+				 * There is a small race condition where wakeIfNecessary was called between
```

Comment:
Can't we achieve the same with a `Semaphore`?

Otherwise LGTM

Reply:
A semaphore is a bit more expensive than an `AtomicBoolean`.

I just tested it and got 18 vs 13 seconds for a billion iterations.